src/modules/uORB/Subscription.hpp

```diff
@@ -39,159 +39,198 @@
-#include <containers/List.hpp>
-#include <systemlib/err.h>
 #include <px4_defines.h>
 
+#include "uORBDeviceNode.hpp"
+#include "uORBManager.hpp"
+#include "uORBUtils.hpp"
+
 namespace uORB
 {
 
-/**
- * Base subscription wrapper class, used in list traversal
- * of various subscriptions.
- */
-class __EXPORT SubscriptionBase
+// Base subscription wrapper class
+class Subscription
 {
 public:
+
 	/**
 	 * Constructor
 	 *
-	 * @param meta The uORB metadata (usually from the ORB_ID()
-	 * 	macro) for the topic.
-	 * @param interval  The minimum interval in milliseconds
-	 * 	between updates
+	 * @param meta The uORB metadata (usually from the ORB_ID() macro) for the topic.
 	 * @param instance The instance for multi sub.
 	 */
-	SubscriptionBase(const struct orb_metadata *meta, unsigned interval = 0, unsigned instance = 0);
-	virtual ~SubscriptionBase();
+	Subscription(const orb_metadata *meta, uint8_t instance = 0) : _meta(meta), _instance(instance)
+	{
+		init();
+	}
 
-	// no copy, assignment, move, move assignment
-	SubscriptionBase(const SubscriptionBase &) = delete;
-	SubscriptionBase &operator=(const SubscriptionBase &) = delete;
-	SubscriptionBase(SubscriptionBase &&) = delete;
-	SubscriptionBase &operator=(SubscriptionBase &&) = delete;
+	virtual ~Subscription() { unsubscribe(); }
+
+	bool init();
+	bool forceInit();
+
+	bool valid() const { return _node != nullptr; }
+	bool published() { return valid() ? _node->is_published() : init(); }
 
 	/**
 	 * Check if there is a new update.
 	 * */
-	bool updated();
+	virtual bool updated() { return published() ? (_node->published_message_count() != _last_generation) : false; }
 
 	/**
 	 * Update the struct
 	 * @param data The uORB message struct we are updating.
 	 */
-	bool update(void *data);
+	virtual bool update(void *dst) { return updated() ? copy(dst) : false; }
 
-	int getHandle() const { return _handle; }
+	/**
+	 * Check if subscription updated based on timestamp.
+	 *
+	 * @return true only if topic was updated based on a timestamp and
+	 * copied to buffer successfully.
+	 * If topic was not updated since last check it will return false but
+	 * still copy the data.
+	 * If no data available data buffer will be filled with zeros.
+	 */
+	bool update(uint64_t *time, void *dst);
+
+	/**
+	 * Copy the struct
+	 * @param data The uORB message struct we are updating.
+	 */
+	bool copy(void *dst) { return published() ? _node->copy(dst, _last_generation) : false; }
 
-	const orb_metadata *getMeta() const { return _meta; }
+	hrt_abstime	last_update() { return published() ? _node->last_update() : 0; }
 
-	unsigned getInstance() const { return _instance; }
+	uint8_t		get_instance() const { return _instance; }
+	orb_id_t	get_topic() const { return _meta; }
 
 protected:
-	const struct orb_metadata *_meta;
 
-	unsigned _instance;
+	bool subscribe();
+	void unsubscribe();
 
-	int _handle{-1};
+	DeviceNode		*_node{nullptr};
+	const orb_metadata	*_meta{nullptr};
+	unsigned		_last_generation{0};
```

Comment:
It's a bit of a hack to work around the current situation where subscribers create the device node (another few kilobytes of wasted memory). Next I want to change the generated output slightly so that we can maintain a bitset of published topics and afford to check for existence every iteration.

For right now I'll add a comment.